File: coding-interview-patterns/python/patterns/binary_search.py
def search_range(nums,target):
    def lower():
        lo,hi=0,len(nums)
        while lo<hi:
            mid=(lo+hi)//2
            if nums[mid]>=target: hi=mid
            else: lo=mid+1
        return lo
    def upper():
        lo,hi=0,len(nums)
        while lo<hi:
            mid=(lo+hi)//2
            if nums[mid]>target: hi=mid
            else: lo=mid+1
        return lo
    l=lower(); r=upper()-1
    return [l,r] if l<len(nums) and l<=r and nums[l]==target else [-1,-1]

def ship_within_days(weights, days):
    lo,hi=max(weights),sum(weights)
    def can(cap):
        d=1; cur=0
        for w in weights:
            if cur+w>cap: d+=1; cur=0
            cur+=w
        return d<=days
    while lo<hi:
        mid=(lo+hi)//2
        if can(mid): hi=mid
        else: lo=mid+1
    return lo

def min_eating_speed(piles,h):
    lo,hi=1,max(piles)
    def ok(v): return sum((p+v-1)//v for p in piles)<=h
    while lo<hi:
        mid=(lo+hi)//2
        if ok(mid): hi=mid
        else: lo=mid+1
    return lo

File: coding-interview-patterns/scala/src/main/scala/patterns/BinarySearch.scala
package patterns
object BinarySearch {
  def searchRange(nums:Array[Int], target:Int): Array[Int] = {
    def lower:Int = { var lo=0; var hi=nums.length; while (lo<hi){ val mid=(lo+hi)/2; if (nums(mid)>=target) hi=mid else lo=mid+1 }; lo }
    def upper:Int = { var lo=0; var hi=nums.length; while (lo<hi){ val mid=(lo+hi)/2; if (nums(mid)>target) hi=mid else lo=mid+1 }; lo }
    val l=lower; val r=upper-1
    if (l<nums.length && l<=r && nums(l)==target) Array(l,r) else Array(-1,-1)
  }
  def shipWithinDays(weights:Array[Int], days:Int): Int = {
    var lo=weights.max; var hi=weights.sum
    def ok(cap:Int):Boolean = {
      var d=1; var cur=0
      for (w <- weights){ if (cur+w>cap){ d+=1; cur=0 }; cur+=w }; d<=days
    }
    while (lo<hi){ val mid=(lo+hi)/2; if (ok(mid)) hi=mid else lo=mid+1 }
    lo
  }
  def minEatingSpeed(piles:Array[Int], h:Int): Int = {
    var lo=1; var hi=piles.max
    def ok(v:Int)= piles.map(p => (p+v-1)/v).sum <= h
    while (lo<hi){ val mid=(lo+hi)/2; if (ok(mid)) hi=mid else lo=mid+1 }
    lo
  }
}
